app/controllers/tree_display_controller.rb

```diff
@@ -124,162 +108,212 @@ def list
+    update_is_available(tmp_object, instructor_id, node)
+    assignments_func(node, tmp_object) if node_type == "Assignments"
+  end
+
+  def res_node_for_child(tmp_res)
+    res = {}
+    tmp_res.keys.each do |node_type|
+      res[node_type] = []
+      tmp_res[node_type].each do |node|
+        tmp_object = {
+          "nodeinfo" => node,
+          "name" => node.get_name,
+          "type" => node.type
+        }
+        if node_type == 'Courses' || node_type == "Assignments"
+          courses_assignments_obj(node_type, tmp_object, node)
         end
-        res[nodeType] << tmpObject
+        res[node_type] << tmp_object
       end
+    end
+    res
+  end
+
+  def update_fnode_children(fnode, tmp_res)
+    # fnode is the parent node
+    # ch_nodes are childrens
+    ch_nodes = fnode.get_children(nil, nil, session[:user].id, nil, nil)
+    tmp_res[fnode.get_name] = ch_nodes
+  end
 
+  def init_fnode_update_children(params, node, tmp_res)
+    fnode = (params[:reactParams][:nodeType]).constantize.new
+    node.each do |a|
+      fnode[a[0]] = a[1]
     end
+    update_fnode_children(fnode, tmp_res)
+  end
+
+  # for folder nodes
+  def children_node_ng
+    child_nodes = child_nodes_from_params(params[:reactParams][:child_nodes])
+    tmp_res = {}
+    child_nodes.each do |node|
+      init_fnode_update_children(params, node, tmp_res)
 
+      # res = res_node_for_child(tmp_res)
+      # cnode = fnode.get_children("created_at", "desc", 2, nil, nil)
+    end
+    res = res_node_for_child(tmp_res)
     respond_to do |format|
       format.html { render json: res }
     end
   end
 
+  def coursenode?(ta_mappings, node)
+    ta_mappings.each do |ta_mapping|
+      return true if ta_mapping.course_id == node.node_object_id
+    end
+  end
+
+  def assignmentnode?(ta_mappings, node)
+    course_id = Assignment.find(node.node_object_id).course_id
+    ta_mappings.each do |ta_mapping|
+      return true if ta_mapping.course_id == course_id
+    end
+  end
+
   def ta_for_current_course?(node)
     ta_mappings = TaMapping.where(ta_id: session[:user].id)
     if node.type == "CourseNode"
-      ta_mappings.each do |ta_mapping|
-        return true if ta_mapping.course_id == node.node_object_id
-      end
+      return true if coursenode?(ta_mappings, node)
     elsif node.type == "AssignmentNode"
-      course_id = Assignment.find(node.node_object_id).course_id
-      ta_mappings.each do |ta_mapping|
-        return true if ta_mapping.course_id == course_id
-      end
+      return true if assignmentnode?(ta_mappings, node)
     end
     false
   end
 
-  # for child nodes
-  def get_children_node_2_ng
-    childNodes = {}
-    childNodes = if params[:reactParams2][:child_nodes].is_a? String
-                   JSON.parse(params[:reactParams2][:child_nodes])
-                 else
-                   params[:reactParams2][:child_nodes]
-                 end
-    tmpRes = {}
+  def available_condition2?(instructor_id, child)
```

Comment:
could you figure out what does this method do and give it a better name? Thanks,